crates/sound/src/stream.rs

```diff
@@ -251,33 +252,66 @@ impl std::fmt::Debug for Buffer {
     }
 }
 
-impl Buffer {
-    pub fn new(data_descriptor: virtio_queue::Descriptor, message: Arc<IOMessage>) -> Self {
+impl Request {
+    pub fn new(len: usize, message: Arc<IOMessage>) -> Self {
         Self {
+            len,
             pos: 0,
-            data_descriptor,
             message,
+            pos_curr_desc: 0,
+            // payload begins after the header descriptor
+            idx_curr_desc: 1,
         }
     }
 
-    pub fn consume(&self, buf: &mut [u8]) -> Result<u32> {
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
-            .message
-            .desc_chain
-            .memory()
-            .read(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+    pub fn consume(&mut self, buf: &mut [u8]) -> Result<u32> {
+        let mut count = buf.len();
+        let descriptors: Vec<_> = self.message.desc_chain.clone().collect();
+        let mut buf_pos = 0;
+
+        while count > 0 {
+            // payload finishes at index (descriptor.len() - 2)
+            if self.idx_curr_desc > descriptors.len() - 1 {
+                panic!(
+                    "idx_desc: {} > len: {}!",
+                    self.idx_curr_desc,
+                    descriptors.len() - 1
+                );
```

Comment:
I think we can avoid the need to panic here, and instead return an error what do you think.

Reply:
I think you're right. I was not sure what to do in this situation. The panic is because consume() is reading more that it is possible. This should never happen because caller is passing to consume() a buf with the correct size, e.g., request.len() - request.pos(). However, this may not always the case and consume() may read less than the buffer size. I'll avoid panic and instead, log a msg and return.